### ISSUE TITLE
Docker image storage policy on Hal

### ISSUE THREAD

**Issue by lzamparo** [juanperin/mskcc-hpc#2]
Hi,

There have been a couple of threads on the cBio-cluster issue tracker pertaining to docker policy that I'd like to echo here.  It's particularly a big deal now since changes made at docker hub have now [complicated workflows](https://github.com/cBio/cbio-cluster/issues/358) for anyone using docker on Hal.

In the first, @tatarsky is trying to gauge what's the best way to operate docker on the Hal cluster.  Currently, all nodes have their own separate storage space for previously pulled docker images, which are not accessible cluster wide.  So, if I submit a docker job which pulls an image from docker hub into the queue, and it runs on node 1, that image is only stored and accessible to docker on node 1.  If I want to re-run that job later, I have to run `docker pull` again as the first part of my job.

This is a drag, since as of December 15th, the version of docker running on all Hal nodes can no longer pull from docker hub (viz. the second [issue](https://github.com/cBio/cbio-cluster/issues/358) for why this is a non-trivial problem).  

@tatarsky has heroically figured out a way to avoid a cluster-wide OS upgrade, but currently only a couple of nodes have this fix.  If these two nodes could also serve as the host of all pulled docker images that were made available cluster-wide, we could continue to use docker as before while we figure out the least-painful path to an OS or kernel upgrade.
### TL;DR

I propose we make a change to the docker image storage policy, by making one (or both) of the currently available nodes host the docker images in its scratch space, and that this space be made available to all nodes on the network for docker jobs.

**Comment by lzamparo** [comment]
@jchodera: your group are heavy docker users too, what do you think?

**Comment by jchodera** [comment]
I like the idea of an interim solution, but I think it might be easier to just designate a GPFS shared directory (/cbio/shared/docker?). 

We should also discuss with @juanperin a kernel.upgrade path. Maybe the new structural biology nodes are a good candidate for updated CentOS 7.x, with torque attributes set to distinguish CentOS versions?

**Comment by tatarsky** [comment]
Couple of notes. I don't follow this Git and didn't even know it existed so I'm just warning I don't have it monitored and saw this from the road. 

Please note I am traveling and not overly available until the 14th when I will be at JHU doing a cluster upgrade so any changes are probably not going to occur until then. 

However I agree generically with the points above and asked in https://github.com/cBio/cbio-cluster/issues/358 if people were ready to give the shared GPFS docker dir a try. I'd rather not turn a compute node into a file server. If that was truly needed because GPFS didn't work I would suggest a head node be used for that role. (Several are coming).

Also it would I think be wiser to do initial tests of that GPFS docker spool on a system not actively running other peoples jobs (as `cpu-6-1` and `cpu-6-2` are busy busy machines) and I will need to prep the new kernel on a system out of torque at the moment to test if it works. I do not have time for that from the road but can attempt it if I get some cycles waiting on installs at JHU. (`gpu-3-8` and `gpu-3-9` are good places to test this as they are connected for now to the new Torque/Moab head and not in the main scheduler path)

I could if desperate attempt the image storage dir change on `cpu-6-1` and `cpu-6-2` from the road but I'd sort of like that stated as desired by @juanperin as something he wants me to do. My support availability if something goes wrong is very limited the next several days. 

Generically I think you should all sit with @juanperin and set requirements for 2016 for Docker. If CentOS 7 is to be included in support, my scope and quote for 2016 would be altered to include that (It does not include it now). Also I feel he should confirm he wishes my time spent on this Docker work compared to his staff. 

Hope that helps and advise if the above makes no sense.

**Comment by jchodera** [comment]
I didn't know this exists as well, and I don't think people are monitoring this.

@lzamparo : Can you migrate discussion to https://github.com/cBio/cbio-cluster/issues

**Comment by juanperin** [comment]
This exists for the HPC user group, so the overall discussion that Lee started having a greater overall implication to all MSK users is useful here in my opinion.

I agree to with the options, and an interim solution will need to be addressed to ensure work can progress.  The shared docker location seems logical, so i’ll address that in the cbio thread.  

Long term docker solutions will be provided.  There are a number of other issues to address first on a much larger scale that this clearly instigates and as Paul pointed out will take some careful planning.  I’ll look for such input as we devise this plan, so thank you Lee and all for your experiences and personnel recommendations.  

Juan

**Comment by lzamparo** [comment]
I'm happy to do so.  Discussion is [migrated here](https://github.com/cBio/cbio-cluster/issues/360)

**Comment by lzamparo** [comment]
@juanperin : should I close this issue until the migrated discussion is resolved?

**Comment by lzamparo** [comment]
Maybe I should close this for tidyness' sake, it seems all parties involved have found a resolution.
